contracts/TurnupSharesV4.sol

```diff
@@ -274,16 +275,33 @@ contract TurnupSharesV4 is Initializable, OwnableUpgradeable {
   // @param sharesSubject The subject of the shares
   // @param amount The amount of shares to buy
   function buyShares(address sharesSubject, uint256 amount) public payable virtual onlyIfSetup {
+    (, uint256 excess) = _buyShares(sharesSubject, amount, msg.value, true);
+    if (excess > 0) _sendFundsBackIfUnused(excess);
+  }
+
+  function _buyShares(
+    address sharesSubject,
+    uint256 amount,
+    uint256 expectedPrice,
+    bool revertOnPriceError
+  ) internal returns (bool, uint256) {
+    if (amount == 0) {
+      if (revertOnPriceError) revert InvalidAmount();
+      else return (false, expectedPrice);
+    }
     uint256 supply = getSupply(sharesSubject);
     // solhint-disable-next-line reason-string
-    if (!(supply > 0 || sharesSubject == _msgSender())) revert OnlyKeysOwnerCanBuyFirstKey();
+    if (supply == 0 && sharesSubject != _msgSender()) revert OnlyKeysOwnerCanBuyFirstKey();
```

Comment:
Before the user buys his first key, others are not allowed to buy it。 there may be a bug here